### DISCUSSION TITLE
5.0.0-beta108 fails to compile SFML on FreeBSD

### DISCUSSION THREAD

**Discussion by PaddyMac** [4ian/GDevelop#2513] [Technical questions about the codebase]
I downloaded the GDevelop source code for 5 beta 108, ran cmake in a build directory, and ran make. The compilation failed while compiling SFML. The thing is, this is apparently a problem with GDevelop because I downloaded the SFML code from git master and compiled it by itself, and it compiled successfully. A cmake option to use an external, system library would be nice. I'm running FreeBSD 13-RC5.

[gdevelop5-beta108.txt](https://github.com/4ian/GDevelop/files/6261935/gdevelop5-beta108.txt)

**Comment by Bouh** [comment]
SFML isn't used since GD4.
GD5 and all beta use Pixi.JS as graphic library.
Did you follow the [right README?](https://github.com/4ian/GDevelop#getting-started)

**Reply by arthuro555** [reply]
For the record, when I tried to compile GDCpp when working on Array and Boolean variables, I couldn't manage to get SFML-audio to compile which made testing it very hard

**Reply by 4ian** [reply]
@arthuro555 would you be able to try just compiling test cases to see if it works?
- `mkdir Binaries/build && cd Binaries/build && ccmake ../..`
- Activate GD_BUILD_TESTS (we should set it to TRUE by default)
- `make -j 4 GDCore_tests GDCpp_tests && Core/GDCore_tests  && GDCpp/GDCpp_tests`

At least tests should pass. I think we should remove the GDCpp runtime target, it's of no use anymore and we might remove SFML entirely by exposing just what is needed (an equivalent for sf::Vector2f and... that's it?).

**Reply by arthuro555** [reply]
Hmmm I have switched machines in the meantime, and now GDCpp tests compile fine but dont run:
![image](https://user-images.githubusercontent.com/19349038/114688670-368cc000-9d15-11eb-9139-29889173c5ee.png)

And GDCore tests refuse to link:
```
[build] [ 88%] Linking CXX executable GDCore_tests.exe
[build] CMakeFiles\GDCore_tests.dir/objects.a(ExpressionCodeGenerator.cpp.obj): In function `ZN2gd26ExpressionParserDiagnostic10GetMessageEv':
[build] c:/code/gdevelop/core/gdcore/events/parsers/ExpressionParser2Node.h:47: undefined reference to `gd::ExpressionParserDiagnostic::noMessage'
[build] collect2.exe: error: ld returned 1 exit status
[build] Core\CMakeFiles\GDCore_tests.dir\build.make:598: recipe for target 'Core/GDCore_tests.exe' failed
```

**Reply by arthuro555** [reply]
It seems I need to make the dlls for GDCore and GDCpp first, but I can't get make to generate those for some reason 🤔

**Reply by arthuro555** [reply]
Apparently all compilations errors are due to this: 
![image](https://user-images.githubusercontent.com/19349038/114691953-7acd8f80-9d18-11eb-888d-2a2797002fc9.png)

**Reply by arthuro555** [reply]
Eh I fixed errors after errors and I am tired of it. The tests simply won't compile without major chnages on my PC with Windows, MinGW GCC 6.3.0, and CMake 3.19.4. I'll try later on WSL to see if I have more luck.

**Reply by 4ian** [reply]
This is a Windows macro getting in our way. 
Try

    #undef GetObject 

at the top of the file.

**Reply by arthuro555** [reply]
Yes I had already figured, but after this came another error then another then another...

**Reply by 4ian** [reply]
Ok, no need to push things much further then. We could configure AppVeyor in the future to run these on Windows so that they are always tested to be working after we fix them.

**Comment by arthuro555** [comment]
I have 0 experience with it so I may be mistaken but wouldn't be docker a good fit for improving developer experience here? IIRC it is useful to create a build environment that will be the same independently of the machine it is run on. This would help (for tests at least) to compile from any platforms without much issues.

**Reply by 4ian** [reply]
I think this is a bit overkilled to maintain a Docker for this (you also pay the overhead of docker running, need port forwarding and several headache that will for sure burn multiple hours of work).

This being said, take a look at subscribing to the beta of GitHub workspace. They basically give you free docker powered machines in the cloud. I've recently tested this and it works well with GDevelop it seems. Though I've just compiled libGD.js, not the native tests

**Reply by 4ian** [reply]
Note that Docker is awesome for things like server side apps, the online build run on this :)
For a dev environment, I would strongly favour to use as much as possible no such abstraction to avoid adding complexity to what should be in theory "just use Node.js" and "just run CMake".

**Reply by arthuro555** [reply]
GDevelop.js never cause any issues weirdly enough 🤔 . Only the direct cmake builds, I guess it could be because it only let me use GCC/G++, not the default msvc compiler (when I try to use msvc cmake complains about it not being compatible with GNU extensions to C++ 11 required by the project)? 
I also tried to build from WSL, it didn't work out, as it doesn't support some libraries required by SFML.

> a look at subscribing to the beta of GitHub workspace.

I have actually been in the beta for quite some time now, but am a bit scared to use it as it will be paid after getting out of beta. I don't think it is viable to have to push half done commits to pull them there and run the tests. I cannot make it my main development environment either as I am not working on GDevelop always online and it won't be free forever. At this point, I better just wait for the CI to run the tests, even if it is suboptimal :/

**Reply by arthuro555** [reply]
I just tested github codespaces, and it doesn't compile there either.
Commands used for compilation:
```
sudo apt-get update
sudo apt-get install libgtk-3-dev cmake p7zip-full libopenal-dev libjpeg-dev libglew-dev libudev-dev libxrandr-dev libsndfile1-dev libglu1-mesa-dev libfreetype6-dev
cd Binaries && mkdir build && cd build
cmake ../.. -DBUILD_TESTS=ON
make GDCore -j6
```

Error: https://pastebin.com/MBc3Tw1f
